Lib/_pyio.py

```diff
@@ -2004,19 +2027,22 @@ class TextIOWrapper(TextIOBase):
     def __init__(self, buffer, encoding=None, errors=None, newline=None,
                  line_buffering=False, write_through=False):
         self._check_newline(newline)
-        if encoding is None:
+        encoding = text_encoding(encoding)
+
+        if encoding == "locale":
             try:
-                encoding = os.device_encoding(buffer.fileno())
+                encoding = os.device_encoding(buffer.fileno()) or "locale"
             except (AttributeError, UnsupportedOperation):
                 pass
-            if encoding is None:
-                try:
-                    import locale
-                except ImportError:
-                    # Importing locale may fail if Python is being built
-                    encoding = "ascii"
-                else:
-                    encoding = locale.getpreferredencoding(False)
+
+        if encoding == "locale":
+            try:
+                import locale
+            except ImportError:
+                # Importing locale may fail if Python is being built
+                encoding = "utf-8"
+            else:
+                encoding = locale.getpreferredencoding(False)
```

Comment:
note: io.TextIOWrapper now uses _locale.get_locale_encoding(). locale.getpreferredencoding(False) is correct and I prefer to keep it since _locale is an implementation detail, whereas _pyio should work on an Python implementation.

Reply:
This is slightly off-topic, but if we have public `locale.get_locale_encoding()` API, I am considering it returns actual locale encoding even if UTF-8 mode is enabled.

If we enable UTF-8 mode by default on all platform at some day, there is no public API to get actual locale encoding.

Reply:
PEP 597 is accepted as `encoding="locale"` is same to `encoding=None`.
But if we enable UTF-8 mode by default at some point, I prefer:

* `encoding=None` : Uses UTF-8 if UTF-8 mode is enabled, otherwise locale encoding.
* `encoding="locale"` : Uses locale encoding regardless UTF-8 mode.

I think we should discuss about it at the language summit. But I am not sure about I can attend it...

Reply:
_locale.get_locale_encoding() calls _Py_GetLocaleEncoding() which returns UTF-8 if the Python UTF-8 Mode is enabled.

Maybe the function could have a flag: please don't lie to me and return the current locale encoding ;-)

Or we could add a function to get the *current* locale encoding: **locale.get_current_locale_encoding()**.

This one would ignore the UTF-8 Mode and call nl_langinfo(CODESET). There are APIs to use the *current* locale encoding: PyUnicode_EncodeLocale/PyUnicode_DecodeLocale and _Py_EncodeLocaleEx/_Py_DecodeLocaleEx with current_locale=1. You can see which functions use it:

* decode tm_zone field of localtime_r() and gmtime()
* decode tzname[0] and tzname[1] strings
* decode setlocale() result
* decode some localeconv() fields (this function requires to switch to different locale encoding, it's bad!)
* decode nl_langinfo() result
* decode gettext(), dgettext(), dcgettext(), textdomain(), bindtextdomain(), bind_textdomain_codeset() result
* decode strerror() and dlerror() result
* encode/decode in the readline module
* encode format string for strftime() in time.strftime() (only used on Windows, Unix provides wcsftime) and then decode strftime() result

Reply:
> encoding="locale" : Uses locale encoding regardless UTF-8 mode.

Currently, open(encoding=None) doesn't work like that. For example, on macOS, Android and VxWorks, it always use UTF-8. And if the UTF-8 Mode is used, UTF-8 is used.

In the PEP 597, I read the encoding="locale" is the same than encoding=None but don't emit an EncodingWarning. Where the PEP 597 changes the chosen encoding for encoding=None case? The PEP says "locale encoding" without specifying exactly what it is. In Python, it means different things depending on the context. There is subtle difference the **current** locale encoding and "the locale encoding". I agree that it needs some clarification :-)

While we discuss encodings, I never understood why open() gets the current locale encoding from nl_langinfo(CODESET), encoding which can change at runtime while Python is running. For example, if thread A calls open(filename, encoding=None), thread B calls locale.localeconv(), and the LC_MONETARY locale uses a different encoding than the LC_CTYPE locale, thread A can get the LC_MONETARY encoding because of how locale.localeconv() is currently implemented: it changes temporarily LC_CTYPE to LC_MONETARY to decode the monetary fields of localeconv() result.

I would prefer that Python uses the same encoding for the whole lifetime of the process, since the beginning until the end. The Python filesystem encoding is a good choice for that. It's the same than locale.getpreferredencoding(False) (currently used by open() and friends), but becomes different if the LC_CTYPE is changed (temporarily or permanently).

Reply:
I created https://bugs.python.org/issue43552 "Add  locale.get_locale_encoding() and locale.get_current_locale_encoding()".

Reply:
I am not talking about this pull request or PEP 597. `encoding="locale"` and `encoding=None` use exact same encoding in this pull request and PEP 597.

I am talking about "if we enable UTF-8 mode by default in the future". So this is off-topic. And this is a subtopic of larger topic: how we change the default text encoding?

If we just change the default text encoding, there is no problem. `encoding="locale"` uses locale encoding anyway so it is backward compatible.

On the other hand, if we make UTF-8 mode default, it may break `encoding="locale"` too. Users using `encoding="locale"` may need to disable UTF-8 mode even though they follow the recommendation "use `encoding="locale"` over `encoding=None`".


> I would prefer that Python uses the same encoding for the whole lifetime of the process, since the beginning until the end. The Python filesystem encoding is a good choice for that. It's the same than locale.getpreferredencoding(False) (currently used by open() and friends), but becomes different if the LC_CTYPE is changed (temporarily or permanently).

filesystem encoding is different to `locale.getpreferredencoding(False)` on Windows, unless PYTHONLEGACYWINDOWSFSENCODING is used.

Does we need yet another encoding? Maybe, `sys.locale_encoding` or `os.locale_encoding`.

Reply:
First I didn't understand that you was talking about the step *after* the PEP 597. Ok. If tomorrow Python uses UTF-8 by default, in this case, open(encoding="locale") should use nl_langinfo(CODESET) encoding, rather than calling locale.getpreferredencoding(False).

By the way, would it be possible to modify io.text_encoding(None) and io.text_encoding("locale") to return the locale encoding, rather than returning "locale"?

I would also prefer moving text_encoding() call from open() into TextIOWrapper constructor. What do you think?

Reply:
The idea is that if you want to reimplement the same behvior than TextIOWrapper without using the io module, you can just call text_encoding() and you get directly the same encoding, rather than having to reimplement the logic.

Oh, one issue is that currently text_encoding() doesn't take a file descriptor parameter.

Maybe another helper function is needed? Or maybe my idea is not worth it.

Reply:
> By the way, would it be possible to modify io.text_encoding(None) and io.text_encoding("locale") to return the locale encoding, rather than returning "locale"?

> Oh, one issue is that currently text_encoding() doesn't take a file descriptor parameter.

Yes, you are right.

> I would also prefer moving text_encoding() call from open() into TextIOWrapper constructor. What do you think?

Then, what is *stacklevel* for the EncodingWarning?

> Maybe another helper function is needed? Or maybe my idea is not worth it.

I think it is not worth enough. `os.device_encoding(fd)` is not so important. It is just locale encoding on Unix. TextIOWrapper always use UTF-8 for WindowsConsoleIO. So `os.device_encoding(fd)` is useful only for PYTHONLEGACYWINDOWSSTDIO.

So I think implementing same behavior is a bad idea. It is too complex and benefit is too small.